### PR TITLE
Add golangci-lint to github workflow

### PR DESCRIPTION
**Pull request by omerap12**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
Adds golangci-lint to verify code quality along with golint (ref: https://github.com/kubernetes/autoscaler/pull/7542#issuecomment-2514003952)
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

**Comment by omerap12** [comment]
I've added this action: [golangci-lint-action v6](https://github.com/golangci/golangci-lint-action/tree/v6/). However, I’m not happy with the duplicated code—perhaps someone can suggest a better approach.

One possibility is to create a separate GitHub workflow file to run the linter in parallel using a matrix configuration. 
example:
```yaml
lint:
  runs-on: ubuntu-latest
  strategy:
    matrix:
      dir: [vertical-pod-autoscaler, cluster-autoscaler, addon-resizer, balancer, multidimensional-pod-autoscaler]
  steps:
    - name: Set up Go
      uses: actions/setup-go@v2
      with:
        go-version: '1.22.2'

    - uses: actions/checkout@v2
      with:
        path: ${{ env.GOPATH }}/src/k8s.io/autoscaler

    - name: golangci-lint
      uses: golangci/golangci-lint-action@v6
      with:
        working-directory: ${{ env.GOPATH }}/src/k8s.io/autoscaler/${{ matrix.dir }}
        version: v1.60
```
Alternatively, we could skip GitHub Actions entirely and integrate this manually into the existing script: [verify-golint.sh](https://github.com/kubernetes/autoscaler/blob/master/hack/verify-golint.sh).

Any thoughts or suggestions?
cc @adrianmoisey @voelzmo

**Comment by raywainman** [comment]
FWIW I think the way you have it here is OK.

Moving these up into GitHub Actions seems like the right path forward to me, the UX is also slightly better that way with a nice breakdown in each PR for each action.

So overall direction here LGTM.

**Comment by adrianmoisey** [comment]
I'm also fine with the duplication, it isn't that big of a deal

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -37,6 +37,27 @@ jobs:
         run: hack/verify-all.sh -v
         env:
           GO111MODULE: auto
+      
+      - name: golangci-lint
```
Comment (adrianmoisey) [inline .github/workflows/ci.yaml:41]:
```suggestion
      - name: golangci-lint - balancer
```

Comment (adrianmoisey) [reply .github/workflows/ci.yaml:41]: The Github Actions output isn't clear about which job failed:
image

**Comment by adrianmoisey** [comment]
Could you enable checks? I want to see what the annotations look like: https://github.com/golangci/golangci-lint-action?tab=readme-ov-file#annotations

File: vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator.go [hunk 2]
```diff
@@ -193,9 +193,9 @@ func (calc *UpdatePriorityCalculator) GetProcessedRecommendationTargets(r *vpa_t
...
 }
 
-func parseVpaObservedContainers(pod *apiv1.Pod) (bool, sets.Set[string]) {
+func parseVpaObservedContainers(pod *apiv1.Pod) (bool, sets.String) {
 	observedContainers, hasObservedContainers := pod.GetAnnotations()[annotations.VpaObservedContainersLabel]
-	vpaContainerSet := sets.New[string]()
+	vpaContainerSet := sets.NewString()
 	if hasObservedContainers {
 		if containers, err := annotations.ParseVpaObservedContainersValue(observedContainers); err != nil {
 			klog.ErrorS(err, "VPA annotation failed to parse", "pod", klog.KObj(pod), "annotation", observedContainers)
```
Comment (omerap12) [inline vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator.go:201]: I have changed this just for testing purposes . Ill will revert this before merge

**Comment by omerap12** [comment]
> Could you enable checks? I want to see what the annotations look like: https://github.com/golangci/golangci-lint-action?tab=readme-ov-file#annotations

done

**Comment by adrianmoisey** [comment]
I tested out annotations in https://github.com/kubernetes/autoscaler/pull/7558/files, I like them

**Comment by adrianmoisey** [comment]
Hmmm... balancer is failing.
May be this needs to be rolled out project at a time?

**Comment by omerap12** [comment]
@adrianmoisey , added.
I will raise another PR to fix all those golang-lint fixes.

**Comment by omerap12** [comment]
> Hmmm... balancer is failing. May be this needs to be rolled out project at a time?



> Hmmm... balancer is failing. May be this needs to be rolled out project at a time?

GitHub Actions PRs can be hard to merge, so I tried to do everything in one PR. I don’t mind splitting it into a couple of PRs if that works better. @raywainman, what do you think?

**Comment by adrianmoisey** [comment]
> I don’t mind splitting it into a couple of PRs if that works better.

For approval sake, it may make sense to make a PR per project, and once they are merged, add the GitHub Action

**Comment by omerap12** [comment]
So I'm closing this PR for now

**Pull request closed without merging**